### ISSUE TITLE
Fine-tuning with custom datasets?

### ISSUE THREAD

**Issue by sukumar-varma** [faustomorales/keras-ocr#3]
Hi, is it possible for me to fine-tune the model with a custom dataset after training it on syth dataset?

**Comment by faustomorales** [comment]
Hi @sukumar-varma -- Sure, both the detector and recognizer can be fine tune with a custom dataset. For the recognizer, you just need to get crops of lines of text along with string representations of what is in the image. For the detector, character-level annotations are required right now. Could you say a little more about your use case so I can give you a more precise answer?

**Comment by sukumar-varma** [comment]
I am planning to train it on handwritten characters to see if it can segment characters. So I was initially planning on training it on a synth dataset of handwritten fonts and then fine tune it on my dataset which have character level annotations. But I couldn't find any steps in the docs on fine-tuning the model. I'm really sorry if I've missed something from the docs.

**Comment by faustomorales** [comment]
I've added more documentation on this today.

- For an example on fine tuning the recognizer with a custom dataset, please see [this page](https://keras-ocr.readthedocs.io/en/latest/examples/fine_tuning_recognizer.html), which I wrote today. Note that it uses a dataset where each image file is cropped to just one line of text. If your dataset is not cropped but you have the bounding boxes, you can use it in conjunction with `keras_ocr.tools.warpBox` to get just the line of text from the image.
- Fine tuning the detector is a little more complex because of the character level annotations. I'm not aware of a real-world dataset with this level of granularity. I've put some placeholder notes on [this page](https://keras-ocr.readthedocs.io/en/latest/examples/fine_tuning_detector.html).

As I was writing the above, I realized I could make the API a little clearer so I pushed some changes and bumped the version up today.

- The `keras_ocr.Detector.get_batch_generator` now expects an image generator that provides `(image, lines)` tuples. See the output of [keras_ocr.data_generation.get_image_generator](https://keras-ocr.readthedocs.io/en/latest/api.html#keras_ocr.data_generation.get_image_generator) for an explanation of what that means. Please ask questions here if it doesn't make sense. I also appreciate any PRs for making the documentation clearer!
- The `keras_ocr.Recognizer.get_batch_generator` now expects an image generator that provides simple `(image, text)` tuples. This is much more straightforward and demonstrates that the recognizer doesn't do anything with the character-level bounding boxes.

Finally, for the end-to-end synthetic dataset example and the recognizer fine tuning example, I've added links to a couple of Google Colab notebooks so people can experiment as quickly as possible. For convenience, here are the two notebooks.

- [End-to-end training with synthetic data](https://colab.research.google.com/drive/19dGKong-LraUG3wYlJuPCquemJ13NN8R)
- [Fine-tuning recognizer with real data](https://colab.research.google.com/drive/1GeGA_bvh1TCgYNJSkWTWt00CEwaD_Ocl)

**Comment by sukumar-varma** [comment]
Wow, thanks for the updates and the examples! Is it possible to fine-tune the detector with ICDAR 13's (Task 2.2 of focused scene text segmentation) character annotations? If yes, how can I go about doing it? And is it possible to get character bounding boxes from the detector?

**Comment by faustomorales** [comment]
Good catch -- yes, one can fine tune on the ICDAR 13 focused scene text dataset. I can't commit to write up the example right now, but the main idea is to parse the text file labels and generate `(image, lines)` tuples. The [tasks page](https://rrc.cvc.uab.es/?ch=2&com=tasks) for the task explains the file format and how to get the bounding box coordinates. 

If you build a generator that creates the `(image, lines)` tuples from the image and label files, you can pass that to `keras_ocr.detection.Detector.get_batch_generator`. If you take a first pass at it, I'm happy to provide some feedback. I should add that the existing detector has already been trained on ICDAR 13 (see the [main repository for CRAFT](https://github.com/clovaai/CRAFT-pytorch)).

On your second question, yes, it is possible to get bounding boxes from the detector and this forms the basis for the "pseudo-ground truth" approach, which I have not implemented. The concept is documented in section 3.2.2 of [the paper](https://arxiv.org/pdf/1904.01941.pdf).

**Comment by faustomorales** [comment]
There is now a tutorial for [fine-tuning the detector](https://keras-ocr.readthedocs.io/en/latest/examples/fine_tuning_detector.html) using ICDAR 13.

**Comment by faustomorales** [comment]
Closing this issue since there are now a few examples for training with different datasets. Please comment back to re-open if you find more gaps. Thanks!

**Comment by sukumar-varma** [comment]
Thank you very much for these wonderful examples!
